tools/copyblocks/main.go

```diff
@@ -399,35 +320,34 @@ func isAllowedUser(enabled map[string]struct{}, disabled map[string]struct{}, te
 }
 
 // This method copies files within single TSDB block to a destination bucket.
-func copySingleBlock(ctx context.Context, tenantID string, blockID ulid.ULID, markers blockMarkers, srcBkt bucket, copyFunc CopyFunc) error {
-	paths, err := srcBkt.ListPrefix(ctx, tenantID+delim+blockID.String(), true)
+func copySingleBlock(ctx context.Context, tenantID string, blockID ulid.ULID, markers blockMarkers, srcBkt objtools.Bucket, copyFunc objtools.CopyFunc) error {
+	result, err := srcBkt.List(ctx, objtools.ListOptions{
+		Prefix:    tenantID + objtools.Delim + blockID.String(),
+		Recursive: true,
+	})
 	if err != nil {
 		return errors.Wrapf(err, "copySingleBlock: failed to list block files for %v/%v", tenantID, blockID.String())
 	}
+	paths := result.ToNames()
 
 	// Reorder paths, move meta.json at the end. We want to upload meta.json as last file, because it signals to Mimir that
 	// block upload has finished.
 	for ix := 0; ix < len(paths); ix++ {
-		if paths[ix] == block.MetaFilename && ix < len(paths)-1 {
+		if strings.HasSuffix(paths[ix], block.MetaFilename) && ix < len(paths)-1 {
```

Comment:
This seems like the added `meta.json` could have a bug in regards to what has a prefix and what doesn't. I'll double check this and do a non dry-run after to confirm.

Reply:
Looked like there were two bugs here actually, the prefix bug and a double counting bug. Rewrote it to simplify it and did a test run that went well.